ServerCore/Pages/Submissions/Index.cshtml.cs

```diff
@@ -59,17 +90,129 @@ public async Task<IActionResult> OnPostAsync(int puzzleId, int teamId)
+                _context,
+                Event,
+                puzzle,
+                team)).FirstOrDefaultAsync();
+
+            // Note: These submissions are not guaranteed to be sorted, but
+            // they should be entered into the database in-order.
+            Submissions = await _context.Submissions.Where(
+                (s) => s.Team != null &&
+                       s.Team.ID == teamId &&
+                       s.Puzzle != null &&
+                       s.Puzzle.ID == puzzleId)
+                .ToListAsync();
+        }
+
+        /// <summary>
+        ///     Computes whether a team should be locked out from submitting to this puzzle
+        ///     and returns for how long the lockout should be.
+        /// </summary>
+        /// <param name="ev"></param>
+        /// <param name="submissions">Expects submissions in chronological order</param>
+        /// <param name="puzzleState"></param>
+        /// <returns>
+        ///     The number of minutes the team should be locked out of the puzzle or null
+        ///     if the team should not be locked out.
+        /// </returns>
+        private static double? ComputeLockoutTime(
+            Event ev,
+            IList<Submission> submissions,
+            PuzzleStatePerTeam puzzleState)
+        {
+            uint consecutiveWrongSubmissions = 0;
+            bool shouldLockout = false;
+
+            /**
+             * Count the number of submissions in the past N minutes where N is the LockoutSpamDuration
+             * set for the event. If that count exceeds the LockoutSpamCount set for the event, then
+             * the team should be locked out of that puzzle.
+             */
+
+            foreach (Submission s in submissions.Reverse())
+            {
+                // Do not increment on partials
+                if (s.Response != null)
+                {
+                    continue;
+                }
+
+                if (s.TimeSubmitted.AddMinutes(ev.LockoutSpamDuration).CompareTo(
+                        DateTime.UtcNow) < 0)
+                {
+                    break;
+                }
+
+                if (++consecutiveWrongSubmissions >= ev.LockoutSpamCount)
+                {
+                    shouldLockout = true;
+                    break;
+                }
+            }
+
+            if (shouldLockout)
+            {
+                /**
+                 * If this is the team's first lockout for the puzzle or If enough times has passed
+                 * since the team's last lockout, we give them the minimum lockout duration, 1 minute.
+                 */
+                if (puzzleState.LockoutTime == null ||
+                    (DateTime.UtcNow - puzzleState.LockoutTime)?.TotalMinutes >= ev.LockoutForgivenessTime)
+                {
+                    return 1.0;
+                }
+
+                return puzzleState.LockoutStage * ev.LockoutDurationMultiplier;
```

Comment:
The way I read this, the lockout increases forever even if I put a puzzle away for hours and then come back to it. It should not do that, the multiplier should cool down. Consider borrowing some tricks from the lockout computation in the code snippet I provided and removing the LockoutStage property from PuzzleStatePerTeam - though I agree that storing the lockout expiry time in the state is a good performance win.